helper/testhelpers/pluginhelpers/pluginhelpers.go

```diff
@@ -0,0 +1,132 @@
+// Package pluginhelpers contains testhelpers that don't depend on package
+// vault, and thus can be used within vault (as well as elsewhere.)
+package pluginhelpers
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"sync"
+
+	"github.com/hashicorp/vault/sdk/helper/consts"
+	"github.com/mitchellh/go-testing-interface"
+)
+
+var (
+	testPluginCacheLock sync.Mutex
+	testPluginCache     = map[string][]byte{}
+)
+
+type TestPlugin struct {
+	Name     string
+	Typ      consts.PluginType
+	Version  string
+	FileName string
+	Sha256   string
+}
+
+func GetPlugin(t testing.T, typ consts.PluginType) (string, string, string, string) {
+	t.Helper()
+	var pluginName string
+	var pluginType string
+	var pluginMain string
+	var pluginVersionLocation string
+
+	switch typ {
+	case consts.PluginTypeCredential:
+		pluginType = "approle"
+		pluginName = "vault-plugin-auth-" + pluginType
+		pluginMain = filepath.Join("builtin", "credential", pluginType, "cmd", pluginType, "main.go")
+		pluginVersionLocation = fmt.Sprintf("github.com/hashicorp/vault/builtin/credential/%s.ReportedVersion", pluginType)
+	case consts.PluginTypeSecrets:
+		pluginType = "consul"
+		pluginName = "vault-plugin-secrets-" + pluginType
+		pluginMain = filepath.Join("builtin", "logical", pluginType, "cmd", pluginType, "main.go")
+		pluginVersionLocation = fmt.Sprintf("github.com/hashicorp/vault/builtin/logical/%s.ReportedVersion", pluginType)
+	case consts.PluginTypeDatabase:
+		pluginType = "postgresql"
+		pluginName = "vault-plugin-database-" + pluginType
+		pluginMain = filepath.Join("plugins", "database", pluginType, fmt.Sprintf("%s-database-plugin", pluginType), "main.go")
+		pluginVersionLocation = fmt.Sprintf("github.com/hashicorp/vault/plugins/database/%s.ReportedVersion", pluginType)
+	default:
+		t.Fatal(typ.String())
+	}
+	return pluginName, pluginType, pluginMain, pluginVersionLocation
+}
+
+// to mount a plugin, we need a working binary plugin, so we compile one here.
+// pluginVersion is used to override the plugin's self-reported version
+func CompilePlugin(t testing.T, typ consts.PluginType, pluginVersion string, pluginDir string) TestPlugin {
+	t.Helper()
+
+	pluginName, pluginType, pluginMain, pluginVersionLocation := GetPlugin(t, typ)
+
+	testPluginCacheLock.Lock()
+	defer testPluginCacheLock.Unlock()
+
+	var pluginBytes []byte
+
+	dir := ""
+	// detect if we are in the "vault/" or the root directory and compensate
+	if _, err := os.Stat("builtin"); os.IsNotExist(err) {
```

Comment:
optional: this might need to be modified a little if people are going to use this function outside of the root and `vault/` directories. (Maybe putting in a while loop?) This should only impact running tests inside a directory directly with `go test .`, for example, so maybe is not that big of a deal.

Reply:
I think this may need to be modified since Go changes the working directory to the package's path.

Reply:
done in https://github.com/hashicorp/vault/pull/18952/commits/d8f016ac928661348164fa135b0fc0725d3f468a